README.md
# MMM-ImagesPhotos

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror). It will show photos from a directory.

This module reads the images from the _uploads_ directory inside the module.
**Directory**: `~/MagicMirror/modules/MMM-ImagesPhotos/uploads`

## Installation

1. Clone this repository inside your MagicMirror's `modules` folder

```bash
cd ~/MagicMirror/modules
git clone https://github.com/roramirez/MMM-ImagesPhotos
cd MMM-ImagesPhotos
npm install
```

## How it looks

![Demo](.github/animate.gif)

## Config

The entry in `config.js` can include the following options:

<!-- prettier-ignore-start -->
| Option             | Description
|--------------------|-----------
| `opacity`          | The opacity of the image.<br><br>**Type:** `double`<br>Default 0.9
| `animationSpeed`   | How long the fade out and fade in of photos should take.<br><br>**Type:** `int`<br>Default 500
| `updateInterval`   | How long before loading a new image.<br><br>**Type:** `int`(milliseconds) <br>Default 5000 milliseconds
| `getInterval`      | Interval value to get new images from directory.<br><br>**Type:** `int`(milliseconds) <br>Default 60000 milliseconds
| `sequential`       | true or false, whether to process the images randomly(default) or sequentially<br>Default false
| only when position is `NOT` fullscreen_below or fullscreen_above|
| `maxWidth`         | Value for maximum width. Optional, possible values: absolute (e.g. "700px") or relative ("50%") <br> Default 100%
| `maxHeight`        | Value for maximum height. Optional, possible values: absolute (e.g. "400px") or relative ("70%") <br> Default 100%
|only when position `IS` fullscreen_below or fullscreen_above 
| `backgroundColor`  | Value for color used to fill around the image if not fullscreen,  can be #xxyyzz, like #808080 (grey),<br> if fill is true, the backgroundColor setting is ignored<br>Default 'black'
| `fill`             | true or false,  instead of color use a blurred copy of the image to fill around the image, <br>Default false.
| `blur`             | the size of the pixel blur of the background fill, <br>Default 8


Here is an example of an entry in `config.js`:

not fullscreen

```js
{
 module: "MMM-ImagesPhotos",
 position: "middle_center",
 config: {
  opacity: 0.9,
  animationSpeed: 500,
  updateInterval: 5000,
  maxHeight: "500px",
  maxWidth:"500px",
  sequential: false  // process the image list randomly
 }
},
```

fullscreen

```js
{
	module: "MMM-ImagesPhotos",
	position: "fullscreen_below",
	config: {
		opacity: 0.9,
		animationSpeed: 500,
		updateInterval: 5000,
		backgroundColor: 'grey',  // not used if fill is true
		fill: false,   // fill around image with blurred copy of image
		blur: 10,   // only used if fill is true
		sequential: false  // process the image list randomly
	}
},
```


CHANGELOG.md
# Changelog

## [0.2.0] - 2023-12-19

- Add support for nested image folders, and support for sequential image vs random
- Add support for multiple instances
- Add back support for not full screen
- Add keywords to package.json
- Fix the installation instructions
- Fix css for not full screen too
- Fix missing timer in not full screen
- Fix typos
- Replace XMLHttpRequest by fetch
- Replace grunt by node scripts
- Remove unused request and url packages
- Lint code, json and markdown files

## [0.1.1] - 2017-03-13

- Update README documentation

## [0.1.0] - 2017-02-05

First public release


node_helper.js
/*
 * MagicMirror²
 * Node Helper: MMM-ImagesPhotos
 *
 * By Rodrigo Ramìrez Norambuena https://rodrigoramirez.com
 * MIT Licensed.
 */

const express = require("express");
const Log = require("logger");
const NodeHelper = require("node_helper");
const path = require("path");
const fs = require("fs");
const mime = require("mime-types");

module.exports = NodeHelper.create({
  // Override start method.
  config: {},
  path_images: {},
  start() {
    Log.log(`Starting node helper for: ${this.name}`);
  },

  setConfig(id) {
    if (this.config[id].debug) {
      Log.log(`setconfig path=${id}`);
    }
    this.path_images[id] = path.resolve(
      global.root_path,
      "modules/MMM-ImagesPhotos/uploads",
      this.config[id].path
    );
    if (this.config[id].debug) {
      Log.log(`path for : ${this.name} ${id}= ${this.path_images[id]}`);
    }
  },

  // Override socketNotificationReceived method.
  socketNotificationReceived(notification, payload) {
    if (notification === "CONFIG") {
      Log.log(`Config based debug=${payload.id}`);
      this.config[payload.id] = payload;
      this.setConfig(payload.id);
      this.extraRoutes(payload.id);
      this.sendSocketNotification("READY", payload.id);
    }
  },

  /*
   * Create routes for module manager.
   * Recive request and send response
   */
  extraRoutes(id) {
    if (this.config[id].debug) {
      Log.log(`setting path=${id}`);
    }
    const self = this;

    this.expressApp.get(`/MMM-ImagesPhotos/photos/${id}`, (req, res) => {
      self.getPhotosImages(req, res, id);
    });

    this.expressApp.use(
      `/MMM-ImagesPhotos/photo/${id}`,
      express.static(self.path_images[id])
    );
  },

  // Return photos-images by response in JSON format.
  getPhotosImages(req, res, id) {
    if (this.config[id].debug) {
      Log.log(`gpi id=${id}`);
    }
    const directoryImages = this.path_images[id];

    const imgs = this.getFiles(directoryImages, id);
    const imagesPhotos = this.getImages(imgs, id).map((img) => {
      if (this.config[id].debug) {
        Log.log(`${id} have image=${img}`);
      }
      return { url: `/MMM-ImagesPhotos/photo/${id}/${img}` };
    });
    res.send(imagesPhotos);
  },

  // Return array with only images
  getImages(files, id) {
    if (this.config[id].debug) {
      Log.log(`gp id=${id}`);
    }
    const images = [];
    const enabledTypes = ["image/jpeg", "image/png", "image/gif"];

    for (const idx in files) {
      if (idx in files) {
        const type = mime.lookup(files[idx]);
        if (enabledTypes.indexOf(type) >= 0 && type !== false) {
          images.push(files[idx]);
        }
      }
    }

    return images;
  },

  getFiles(filePath, id) {
    if (this.config[id].debug) {
      Log.log(`gf id=${id}`);
    }
    let files = [];
    const folders = [];
    try {
      // Log.log("finding files on path="+path)
      files = fs.readdirSync(filePath).filter((file) => {
        if (this.config[id].debug) {
          Log.log(`found file=${file} on path=${filePath}`);
        }
        if (fs.statSync(`${filePath}/${file}`).isDirectory()) {
          if (this.config[id].debug) {
            Log.log(`${id} saving folder path=${filePath}/${file}`);
          }
          folders.push(`${filePath}/${file}`);
        } else if (!file.startsWith(".")) {
          return file;
        }
      });

      folders.forEach((x) => {
        if (this.config[id].debug) {
          Log.log(`${id} processing for sub folder=${x}`);
        }
        const y = this.getFiles(x, id);
        // Log.log("list"+JSON.stringify(y))
        const worklist = [];
        // Get the number of elements in the base path
        const c = this.path_images[id].split("/").length;
        // Get the rest of the path
        const xpath = x.split("/").slice(c).join("/");
        y.forEach((f) => {
          // If the file doesn't have a path
          if (f.includes("/")) {
            // Use it as is

            worklist.push(f);
          } else {
            // Add it

            worklist.push(`${xpath}/${f}`);
          }
        });
        // Add to the files list
        files = files.concat(worklist);
        if (this.config[id].debug) {
          Log.log(`files after concat=${JSON.stringify(files)}`);
        }
      });
    } catch (exception) {
      Log.log(
        `getfiles unable to access source folder,path=${filePath} will retry, exception=${JSON.stringify(
          exception
        )}`
      );
    }
    if (this.config[id].debug) {
      Log.log(`${id} returning files=${JSON.stringify(files)}`);
    }

    return files;
  }
});


MMM-ImagesPhotos.js
/* global Log MM Module */

/*
 * MagicMirror²
 * Module: MMM-ImagesPhotos
 *
 * By Rodrigo Ramírez Norambuena https://rodrigoramirez.com
 * MIT Licensed.
 */
const ourModuleName = "MMM-ImagesPhotos";

Module.register(ourModuleName, {
  defaults: {
    opacity: 0.9,
    animationSpeed: 500,
    updateInterval: 5000,
    getInterval: 60000,
    maxWidth: "100%",
    maxHeight: "100%",
    retryDelay: 2500,
    path: "",
    fill: false,
    blur: 8,
    sequential: false
  },

  wrapper: null,
  suspended: false,
  timer: null,
  fullscreen: false,

  requiresVersion: "2.24.0", // Required version of MagicMirror

  start() {
    this.photos = [];
    this.loaded = false;
    this.lastPhotoIndex = -1;
    this.config.id = this.identifier;
    this.sendSocketNotification("CONFIG", this.config);
  },
  getStyles() {
    return ["MMM-ImagesPhotos.css"];
  },

  /*
   * Requests new data from api url helper
   */
  async getPhotos() {
    const urlApHelper = `/MMM-ImagesPhotos/photos/${this.identifier}`;
    const self = this;
    let retry = true;

    try {
      const photosResponse = await fetch(urlApHelper);

      if (photosResponse.ok) {
        const photosData = await photosResponse.json();
        self.processPhotos(photosData);
      } else if (photosResponse.status === 401) {
        self.updateDom(self.config.animationSpeed);
        Log.error(self.name, photosResponse.status);
        retry = false;
      } else {
        Log.error(self.name, "Could not load photos.");
      }

      if (!photosResponse.ok) {
        if (retry) {
          self.scheduleUpdate(self.loaded ? -1 : self.config.retryDelay);
        }
      }
    } catch (error) {
      Log.error(self.name, error);
    }
  },
  notificationReceived(notification, payload, sender) {
    // Hook to turn off messages about notiofications, clock once a second
    if (notification === "ALL_MODULES_STARTED") {
      const ourInstances = MM.getModules().withClass(ourModuleName);
      ourInstances.forEach((m) => {
        if (m.data.position.toLowerCase().startsWith("fullscreen")) {
          this.fullscreen = true;
        }
      });
    }
  },
  startTimer() {
    const self = this;
    self.timer = setTimeout(() => {
      // Clear timer value for resume
      self.timer = null;
      if (self.suspended === false) {
        self.updateDom(self.config.animationSpeed);
      }
    }, this.config.updateInterval);
  },

  socketNotificationReceived(notification, payload, source) {
    if (notification === "READY" && payload === this.identifier) {
      // Schedule update timer.
      this.getPhotos();
    }
  },

  /*
   * Schedule next update.
   *
   * argument delay number - Milliseconds before next update.
   *  If empty, this.config.updateInterval is used.
   */
  scheduleUpdate(delay) {
    let nextLoad = this.config.getInterval;
    if (typeof delay !== "undefined" && delay >= 0) {
      nextLoad = delay;
    }

    const self = this;
    setTimeout(() => {
      self.getPhotos();
    }, nextLoad);
  },

  /*
   * Generate a random index for a list of photos.
   *
   * argument photos Array<String> - Array with photos.
   *
   * return Number - Random index.
   */
  randomIndex(photos) {
    if (photos.length === 1) {
      return 0;
    }

    const generate = () => Math.floor(Math.random() * photos.length);

    let photoIndex =
      this.lastPhotoIndex === photos.length - 1 ? 0 : this.lastPhotoIndex + 1;
    if (!this.config.sequential) {
      photoIndex = generate();
    }
    this.lastPhotoIndex = photoIndex;

    return photoIndex;
  },

  /*
   * Retrieve a random photos.
   *
   * return photo Object - A photo.
   */
  randomPhoto() {
    const { photos } = this;
    const index = this.randomIndex(photos);

    return photos[index];
  },

  scaleImage(srcwidth, srcheight, targetwidth, targetheight, fLetterBox) {
    const result = { width: 0, height: 0, fScaleToTargetWidth: true };

    if (
      srcwidth <= 0 ||
      srcheight <= 0 ||
      targetwidth <= 0 ||
      targetheight <= 0
    ) {
      return result;
    }

    // Scale to the target width
    const scaleX1 = targetwidth;
    const scaleY1 = (srcheight * targetwidth) / srcwidth;

    // Scale to the target height
    const scaleX2 = (srcwidth * targetheight) / srcheight;
    const scaleY2 = targetheight;

    // Now figure out which one we should use
    let fScaleOnWidth = scaleX2 > targetwidth;
    if (fScaleOnWidth) {
      fScaleOnWidth = fLetterBox;
    } else {
      fScaleOnWidth = !fLetterBox;
    }

    if (fScaleOnWidth) {
      result.width = Math.floor(scaleX1);
      result.height = Math.floor(scaleY1);
      result.fScaleToTargetWidth = true;
    } else {
      result.width = Math.floor(scaleX2);
      result.height = Math.floor(scaleY2);
      result.fScaleToTargetWidth = false;
    }
    result.targetleft = Math.floor((targetwidth - result.width) / 2);
    result.targettop = Math.floor((targetheight - result.height) / 2);

    return result;
  },

  suspend() {
    this.suspended = true;
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  },
  resume() {
    this.suspended = false;
    if (this.timer === null) {
      this.startTimer();
    }
  },

  getDom() {
    if (this.fullscreen) {
      return this.getDomFS();
    }
    return this.getDomnotFS();
  },

  getDomnotFS() {
    const self = this;
    const wrapper = document.createElement("div");
    const photoImage = this.randomPhoto();

    if (photoImage) {
      const img = document.createElement("img");
      img.src = photoImage.url;
      img.id = "mmm-images-photos";
      img.style.maxWidth = this.config.maxWidth;
      img.style.maxHeight = this.config.maxHeight;
      img.style.opacity = self.config.opacity;
      img.className = "bgimage";
      wrapper.appendChild(img);
      self.startTimer();
    }
    return wrapper;
  },

  getDomFS() {
    const self = this;
    // If wrapper div not yet created
    if (this.wrapper === null) {
      // Create it once, try to reduce image flash on change

      this.wrapper = document.createElement("div");
      this.bk = document.createElement("div");
      this.bk.className = "bgimagefs";
      if (this.config.fill === true) {
        this.bk.style.filter = `blur(${this.config.blur}px)`;
        this.bk.style["-webkit-filter"] = `blur(${this.config.blur}px)`;
      } else {
        this.bk.style.backgroundColor = this.config.backgroundColor;
      }
      this.wrapper.appendChild(this.bk);
      this.fg = document.createElement("div");
      this.wrapper.appendChild(this.fg);
    }
    if (this.photos.length) {
      // Get the size of the margin, if any, we want to be full screen
      const m = window
        .getComputedStyle(document.body, null)
        .getPropertyValue("margin-top");
      // Set the style for the containing div

      this.fg.style.border = "none";
      this.fg.style.margin = "0px";

      const photoImage = this.randomPhoto();
      let img = null;
      if (photoImage) {
        // Create img tag element
        img = document.createElement("img");

        // Set default position, corrected in onload handler
        img.style.left = `${0}px`;
        img.style.top = document.body.clientHeight + parseInt(m, 10) * 2;
        img.style.position = "relative";

        img.src = photoImage.url;
        // Make invisible
        img.style.opacity = 0;
        // Append this image to the div
        this.fg.appendChild(img);

        /*
         * Set the onload event handler
         * The loadurl request will happen when the html is returned to MM and inserted into the dom.
         */
        img.onload = (evt) => {
          // Get the image of the event
          const eventImage = evt.currentTarget;
          Log.log(
            `image loaded=${eventImage.src} size=${eventImage.width}:${eventImage.height}`
          );

          // What's the size of this image and it's parent
          const w = eventImage.width;
          const h = eventImage.height;
          const tw = document.body.clientWidth + parseInt(m, 10) * 2;
          const th = document.body.clientHeight + parseInt(m, 10) * 2;

          // Compute the new size and offsets
          const result = self.scaleImage(w, h, tw, th, true);

          // Adjust the image size
          eventImage.width = result.width;
          eventImage.height = result.height;

          Log.log(`image setting size to ${result.width}:${result.height}`);
          Log.log(
            `image setting top to ${result.targetleft}:${result.targettop}`
          );

          // Adjust the image position
          eventImage.style.left = `${result.targetleft}px`;
          eventImage.style.top = `${result.targettop}px`;

          // If another image was already displayed
          const c = self.fg.childElementCount;
          if (c > 1) {
            for (let i = 0; i < c - 1; i++) {
              // Hide it
              self.fg.firstChild.style.opacity = 0;
              self.fg.firstChild.style.backgroundColor = "rgba(0,0,0,0)";
              // Remove the image element from the div
              self.fg.removeChild(self.fg.firstChild);
            }
          }
          self.fg.firstChild.style.opacity = self.config.opacity;

          self.fg.firstChild.style.transition = "opacity 1.25s";
          if (self.config.fill === true) {
            self.bk.style.backgroundImage = `url(${self.fg.firstChild.src})`;
          }
          self.startTimer();
        };
      }
    }
    return this.wrapper;
  },

  getScripts() {
    return ["MMM-ImagesPhotos.css"];
  },

  processPhotos(data) {
    const self = this;
    this.photos = data;
    if (this.loaded === false) {
      if (this.suspended === false) {
        self.updateDom(self.config.animationSpeed);
      }
    }
    this.loaded = true;
  }
});
